src/test/regress/sql/local_shard_execution.sql

```diff
@@ -1114,6 +1125,127 @@ INSERT INTO event_responses VALUES (16, 666, 'maybe'), (17, 777, 'no')
+
+-- single shard local modification followed by another single shard
+-- local modification sets the 2PC state, but does not use remotely
+BEGIN;
+	INSERT INTO event_responses_no_pkey VALUES (2, 2, 'yes') RETURNING *;
+	INSERT INTO event_responses_no_pkey VALUES (2, 2, 'yes') RETURNING *;
+	SELECT coordinated_transaction_should_use_2PC();
+COMMIT;
+SELECT count(*) FROM pg_dist_transaction;
+SELECT recover_prepared_transactions();
+
+-- two local commands without transaction block set the internal 2PC state
+-- but does not use remotely
+WITH cte_1 AS (INSERT INTO event_responses_no_pkey VALUES (2, 2, 'yes') RETURNING *),
+	 cte_2 AS (INSERT INTO event_responses_no_pkey VALUES (2, 2, 'yes') RETURNING *)
+SELECT bool_or(coordinated_transaction_should_use_2PC()) FROM cte_1, cte_2;
+SELECT count(*) FROM pg_dist_transaction;
+SELECT recover_prepared_transactions();
+
+-- single shard local modification followed by remote single shard
+-- modification uses 2PC
+WITH cte_1 AS (INSERT INTO event_responses_no_pkey VALUES (2, 2, 'yes') RETURNING *),
+	 cte_2 AS (INSERT INTO event_responses_no_pkey VALUES (1, 1, 'yes') RETURNING *)
+SELECT bool_or(coordinated_transaction_should_use_2PC()) FROM cte_1, cte_2;
+SELECT count(*) FROM pg_dist_transaction;
+SELECT recover_prepared_transactions();
+
+-- single shard local modification followed by remote multi-shard
+-- modification uses 2PC
+WITH cte_1 AS (INSERT INTO event_responses_no_pkey VALUES (2, 2, 'yes') RETURNING *),
+	 cte_2 AS (UPDATE event_responses_no_pkey SET user_id = 1000 RETURNING *)
+SELECT bool_or(coordinated_transaction_should_use_2PC()) FROM cte_1, cte_2;
+
+SELECT count(*) FROM pg_dist_transaction;
+SELECT recover_prepared_transactions();
+
+-- a local modification followed by a remote SELECT does not require to
+-- use actual 2PC
+BEGIN;
+	INSERT INTO event_responses_no_pkey VALUES (2, 2, 'yes');
+	SELECT count(*) FROM event_responses_no_pkey;
+COMMIT;
+SELECT count(*) FROM pg_dist_transaction;
+SELECT recover_prepared_transactions();
+
+-- a local SELECT followed by a remote SELECT does not require to
+-- use actual 2PC
+BEGIN;
+	SELECT count(*) FROM event_responses_no_pkey WHERE event_id = 2;
+	SELECT count(*) FROM event_responses_no_pkey;
+COMMIT;
+SELECT count(*) FROM pg_dist_transaction;
+SELECT recover_prepared_transactions();
+
+-- a local SELECT followed by a remote multi-shard UPDATE requires to
+-- use actual 2PC
+BEGIN;
+	SELECT count(*) FROM event_responses_no_pkey WHERE event_id = 2;
+	UPDATE event_responses_no_pkey SET user_id = 1;
+COMMIT;
+SELECT count(*) FROM pg_dist_transaction;
+SELECT recover_prepared_transactions();
+
+-- a local SELECT followed by a remote single-shard UPDATE does not require to
+-- use actual 2PC. This is not different from running a single shard update
+-- inside a tx block
+BEGIN;
+	SELECT count(*) FROM event_responses_no_pkey WHERE event_id = 2;
+	UPDATE event_responses_no_pkey SET user_id = 1 WHERE event_id = 1;
```

Comment:
I'm missing the test case for UPDATE local shard and update remote shard in the same transaction (and one with order of updates swapped). Because those would require a distributed transaction. Maybe these tests already exist somewhere else, if so you can just point me to them.

Reply:
> UPDATE local shard and update remote shard in the same transaction

I guess these are examples of that case: https://github.com/citusdata/citus/pull/5901/files#r849180959
https://github.com/citusdata/citus/pull/5901/files#diff-2fe2b177ed53504afcff0e2c7081de72e509538052cffe5754a59715b9ec305fR2611

And, the reverse is:

 https://github.com/citusdata/citus/pull/5901/files#diff-2fe2b177ed53504afcff0e2c7081de72e509538052cffe5754a59715b9ec305fR2801

Maybe I can add the CTE version of this as well

Reply:
>  Because those would require a distributed transaction.

Actually, you are right, this PR regresses on that front. We should use 2PC, but probably not in some cases. Wait until I fix that

Reply:
I re-ordered the tests, and added few more. Now, almost all the combinations of the following:
- local / distributed
- modify / SELECT
- single shard / multi-shard

Please go over them and see that all make sense to you as well